src/firmware/linux/host/types.rs

```diff
@@ -413,36 +439,48 @@ pub struct SnpSetExtConfig {
     /// to be updated.
     pub config_address: u64,
 
-    /// Address of extended guest request [`CertTable`] or 0 when
-    /// previous certificate should be removed on SNP_SET_EXT_CONFIG.
+    /// Address of extended guest request [`CertTableEntry`] buffer or 0 when
+    /// previous certificate(s) should be removed via SNP_SET_EXT_CONFIG.
     pub certs_address: u64,
 
-    /// Length of the certificates.
-    pub certs_len: u32,
+    /// 4K-page aligned length of the buffer holding certificates to be cached.
+    pub certs_buf: u32,
 }
 
 impl SnpSetExtConfig {
-    pub(crate) fn from_uapi(data: &SnpExtConfig) -> Result<Self, UserApiError> {
-        Ok(Self {
-            certs_address: if data.certs.is_none() {
-                0
-            } else {
-                match CertTable::from_uapi(data.certs.clone().unwrap()) {
-                    Ok(table) => &table as *const CertTable as u64,
-                    Err(e) => return Err(UserApiError::ApiError(e)),
-                }
-            },
-            config_address: if data.config.is_none() {
-                0
-            } else {
-                &data.config.unwrap() as *const SnpConfig as u64
-            },
-            certs_len: if data.certs.is_none() {
-                0
-            } else {
-                data.certs_len
-            },
-        })
+    pub(crate) fn from_uapi(
+        data: &SnpExtConfig,
+        bytes: &mut Vec<u8>,
+    ) -> Result<Self, UserApiError> {
+        // Make sure the buffer is is of sufficient size.
+        if data.certs_buf < bytes.len() as u32 {
+            return Err(UserApiError::ApiError(SnpCertError::BufferOverflow));
+        }
+
+        // Make sure the buffer length is 4K-page aligned.
+        if data.certs_buf > 0 && data.certs_buf as usize % _4K_PAGE != 0 {
+            return Err(UserApiError::ApiError(SnpCertError::PageMisallignment));
+        }
```

Comment:
From my above suggestion about getting rid of `certs_buf` in the user API struct, this check wouldn't be necessary. We could instead calculate the size of the certificates here.

Reply:
The `certs_buf` is only part of the FFI wrapper, and is required to communicate with the ioctl. We have to transform the user api into this.